### ISSUE TITLE
jvmkill should write heapdump to volume service

### ISSUE THREAD

**Issue by nebhale** [cloudfoundry/java-buildpack#439]
When a volume service is mounted and named/labeled/tagged with a reasonable identifier, `jvmkill` should write a heapdump to that location before it kills the JVM.  This project consists of two parts

1.  The `jvmkill` agent should have a new component added in order to optionally write out a heapdump to a specified location when it is configured to do so.  In most cases there will not be volume service mounted, so in most cases, this won't be executed, but configuring it in the `jvmkill` parameters should enable it.  Note that the name will need to both a directory to write in and a prefix/pattern that the name should be.  Parts of the name will be specified by the configuration (e.g. directory and a filename containing app id, instance id, instance index, etc.) but part of the name will be generated (e.g. a timestamp).

2.  The buildpack should detect the appropriate service and configure the `jvmkill` agent with the appropriate values when detected.  It will be responsible for specifying the fixed part of the heap dump name, and enabling the support when appropriate.

**Comment by glyn** [comment]
After discussion with @nebhale, we will use the [HotSpot diagnostic MXBean](https://docs.oracle.com/javase/8/docs/jre/api/management/extension/com/sun/management/HotSpotDiagnosticMXBean.html#dumpHeap-java.lang.String-boolean-). If the MXBean is not available, no heapdump will be produced.

**Comment by michele-mancioppi** [comment]
I think we should have the name of the volume service aligned with https://github.com/cloudfoundry/java-buildpack/pull/436

Also, I am working on S3 upload, which would make perfect sense for jvmkill as well.

I am getting he feeling that the jvmkill agent and the Java Memory Assistant will have such an overlap that the entire "create heapdump when told to" (I.e., JVM crash or the JavaMemoryAssistant invoking a command) could simply be implemented in jvmkill and reused.

**Comment by michele-mancioppi** [comment]
Which would means that the JavaMemoryAssistant becomes the trigger for heapdumps, delegating the heap-dump creation to jvmkill, and it is absolutely fine with me.

**Comment by glyn** [comment]
@michele-mancioppi Please note that `jvmkill` is only driven via the JVMTI resource exhaustion callback, so it isn't able to provide heapdumps on request from the user.

**Comment by michele-mancioppi** [comment]
To my understanding, the plan is to have the script triggered on OutOfMemoryError (specified via '-XX:+HeapDumpOnOutOfMemoryError') execute a 'jvmkill' executable that will do something with the heap dump. If I can invoke the same 'jvmkill' executable from the JavaMemoryAssistant agent, I can then reuse the "do stuff with the heapdump" logic and even contribute some more there.

**Comment by michele-mancioppi** [comment]
Or, let me rephrase: if we factor the "do something with a heapdump" out in a command shipped within the buildpack that can move it to a volume provided via service or upload to S3 or whatever we come up with, we can have consistent logic for heap dumps created due to JVM crash as well as JVM dumps on demand by the user (see, e.g., https://github.com/SAP/cf-cli-java-plugin) and taken automatically depending on predefined conditions (https://github.com/SAP/java-memory-assistant).

**Comment by michele-mancioppi** [comment]
As agreed with @glyn, to align the `jvmkill` behavior with https://github.com/cloudfoundry/java-buildpack/pull/436, the volume should be labelled with `jbp-dumps`.

**Comment by nebhale** [comment]
@michele-mancioppi I'm not committed to that name.  Heap dumps are not just a Java thing, and therefore tagging a service, which is designed to be used broadly across multiple applications and languages as Java Buildpack-specific isn't right.

**Comment by michele-mancioppi** [comment]
Fai enough. I am willing to just adopt any label you guys feel comfortable going with in the interest of having consistency.

**Comment by glyn** [comment]
How about `dumps`?

**Comment by kelapure** [comment]
In the majority of the cases, CF apps are not bound to volume services, it is more likely that the app has access to a S3 store rather than a volume services mount

**Comment by nebhale** [comment]
@kelapure I absolutely disagree with that.  For example, the vast majority of PCF instances (by far the largest installed based of CF) are run in an internet-gapped environment and can't access S3 at all.  I do agree that most applications won't run with a mounted volume service either through.  In fact, I'd be surprised if this feature is used by even 5% of instances in the CF ecosystem.  But that's enough for us to add it.

**Comment by michele-mancioppi** [comment]
Well, there are CF offerings other than Pivotal's. I can say that I see the same as @kelapure

**Comment by nebhale** [comment]
@michele-mancioppi Totally agree that there are other CF offerings, but as a CFF project lead, I have some insight into the number of AIs running on all of those distributions and as a percentage, the number of AIs with access to the internet is a very tiny minority of all CF AIs currently running.

**Comment by kelapure** [comment]
@nebhale  first I agree with you that this feature will be used initially by 5% of instances in the CF ecosystem. Second majority (95%) of the PCF installations run in an environment where ingress and egress is tightly controlled. Typically ingress is only allowed through HTTP or HTTPs via the goRouter and outbound egress is locked down via 1. Application Security Groups and 2. Web Application Firewalls which is why staging of java apps is typically done ONLY using *offline* buildpacks. A true air gapped PCF install is seen in military grade PCF deploys. 

Now of the 5% who are willing to open up write access for their applications it is more likely that they have an _internal_ S3 API compliant object store. The `jvmkill` agent would then have access to write to this internal store. Since PCF 1.10 the support for volume services has created the ability to write these dumps to volume supported by the volume service brokers. 

It is my hypothesis that the support for writing dumps to an internal s3 compliant blob store is more useful than writing to a volume service. Ultimately I don't want to lose sight that this is an issue that helps triaging and debugging of dumps that java developers have to live with. The support for writing to an external endpoint ideally could redirect the stream to either s3 or a volume mount.

**Comment by michele-mancioppi** [comment]
The problem with "direct push to S3" is that the JVM needs anyways a filesystem to dump to.

Often heap dumps are larger than the filesystem free space in the container, which means hat in any case one would need to have an intermediate volume from which to push.

So, in worst case, one can add a watchdog on the intermediate volume and push to S3 from there. Not the easiest or most integrated of approaches, but it is a way to go about it.

**Comment by kelapure** [comment]
@michele-mancioppi makes sense ... thanks for listening guys

**Comment by glyn** [comment]
The `jvmkill` support is provided in https://github.com/cloudfoundry/jvmkill/commit/75fa6e5566c0f21a49b46d4c587ba0dd824bd5b8.
